File: ios/chronometer-ios/DefaultFileSystem.m
/*
 * Appspresso
 *
 * Copyright (c) 2011 KT Hitel Corp.
 *
 * This source is subject to Appspresso license terms.
 * Please see http://appspresso.com/ for more information.
 *
 * THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY
 * KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
 * PARTICULAR PURPOSE.
 */


#import "DefaultFileSystem.h"
#import "DefaultFile.h"
#import "AxRuntimeContext.h"

@implementation DefaultFileSystem

#pragma mark -

-(id)initWithBaseDirectory:(NSString*)path canRead:(BOOL)r canWrite:(BOOL)w {
    if ((self = [super init])) {
        _baseDir = [path retain];
        _isReadable = r;
        _isWritable = w;

        _root = [[DefaultFile alloc] initWithFilePath:_baseDir parent:nil fileSystem:self];
        
        // 쓰기가능한 파일시스템이라면... 베이스디렉토리가 없으면 지금 만들어준다.???
        if(w) {
            NSFileManager *fm = [[NSFileManager alloc]init];
            if (![fm fileExistsAtPath:path]) {
                [fm createDirectoryAtPath:path withIntermediateDirectories:NO attributes:nil error:nil];
            }
            [fm release];
        }
    }
    return self;
}

-(void)dealloc {
    [_root release];
    [_baseDir release];
    [super dealloc];
}

#pragma mark AxFileSystem

-(BOOL)onMount:(NSString*)prefix option:(NSDictionary*)option{
    return TRUE;
}

-(void)onUnmount{
}

-(NSObject<AxFile>*)getRoot{ 
    if (!_isReadable) {
        return nil;
    }
    return [_root retain];
}

-(NSObject<AxFile>*)getFile:(NSString*)path{
    if (!_isReadable || path == nil) {
        return nil;
    }
    if([path length] == 0 || [path isEqualToString:@"/"]) {
        return [_root retain];
    }
    // NOTE: 재귀 호출!!
    NSObject<AxFile>* parent = [self getFile:[path stringByDeletingLastPathComponent]]; 
    return [DefaultFile fileWithFilePath:[_baseDir stringByAppendingPathComponent:path] parent:parent fileSystem:self];
}

-(BOOL)canRead{
    return _isReadable;
}

-(BOOL)canWrite{
    return _isWritable;
}

-(BOOL)isThisValidPathComponent:(NSString*)component {
    if (component == @"." || component == @"..") {
        return NO;
    }
    return YES;
}

-(NSString*)validateVirtualPath:(NSString*)virtualPath {
    // TODO: 앞뒤에 / 등을 포함한 좀 더 정교한 에러 검사
    if ([virtualPath isEqualToString:@"/"]) {
        return nil;
    }
    NSMutableArray *pathComponet = [NSMutableArray arrayWithArray:[virtualPath pathComponents]];
    for (NSString* component in pathComponet) {
        if (![self isThisValidPathComponent:component]){
            return nil;
        }
    }
    if ([virtualPath hasSuffix:@"/"]){
        [pathComponet removeLastObject];
    }
    if ([virtualPath hasPrefix:@"/"]) {
        [pathComponet removeObjectAtIndex:0];
    }
    return [NSString pathWithComponents:pathComponet];
}

-(NSString*)toVirtualPath:(NSString *)nativePath {
    // 지정한 네이티브 경로가
    // 이 파일시스템의 베이스 디렉토리에 아래를 가리키는 경로일때만 변환 가능
    if([nativePath hasPrefix:_baseDir]) {
        return [self validateVirtualPath:[nativePath substringFromIndex:[_baseDir length]+1]];
    }
    return nil;
}

-(NSString*)toNativePath:(NSString *)virtualPath {
    return [_baseDir stringByAppendingPathComponent:[self validateVirtualPath:virtualPath]];
}


#pragma mark -

+(id)fileSystemWithBaseDirectory:(NSString*)path canRead:(BOOL)r canWrite:(BOOL)w {
    return [[[DefaultFileSystem alloc] initWithBaseDirectory:path canRead:r canWrite:w] autorelease];
}

@end

File: android/chronometer-android/src/com/appspresso/api/fs/DefaultFileSystem.java
/*
 * Appspresso
 * 
 * Copyright (c) 2011 KT Hitel Corp.
 * 
 * This source is subject to Appspresso license terms. Please see http://appspresso.com/ for more
 * information.
 * 
 * THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS
 * FOR A PARTICULAR PURPOSE.
 */
package com.appspresso.api.fs;

import java.io.File;
import java.util.Map;
import java.util.Properties;
import java.util.WeakHashMap;
import org.apache.commons.logging.Log;
import com.appspresso.api.AxLog;

/**
 * {@link java.io.File}을 추상화한 앱스프레소 가상 파일시스템({@link AxFileSystem}) 구현체.
 * 
 * @version 1.0
 * @see java.io.File
 * @see DefaultFile
 */
public class DefaultFileSystem implements AxFileSystem {
    private static Log L = AxLog.getLog(AxFileSystem.class);

    private final File rootDirectory;
    private DefaultFile rootAxFile;
    private Map<String, DefaultFile> cache; // key : native absolute path
    private boolean canRead;
    private boolean canWrite;

    public DefaultFileSystem(File rootDirectory) {
        this.rootDirectory = rootDirectory;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onMount(String prefix, Properties options) {
        if (!rootDirectory.isDirectory()) {
            if (L.isDebugEnabled()) {
                L.debug("The root is invalid.");
            }
            return;
        }

        rootAxFile = new DefaultFile(this, rootDirectory, prefix);
        cache = new WeakHashMap<String, DefaultFile>();
        setRead(/* rootAxFile.canRead() */true);
        setWrite(/* rootAxFile.canWrite() */true);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onUnmount() {
        cache.clear();
        setRead(false);
        setWrite(false);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public AxFile getFile(String path) {
        if (!canRead) {
            if (L.isDebugEnabled()) {
                L.debug("This file system is not readable.");
            }
            return null;
        }

        File file = new File((File) rootAxFile.getPeer(), path);
        return createDefaultFile(file);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public AxFile getRoot() {
        if (!canRead) {
            if (L.isDebugEnabled()) {
                L.debug("This file system is not readable.");
            }
            return null;
        }

        return rootAxFile;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean canRead() {
        return canRead;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public boolean canWrite() {
        return canWrite;
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toNativePath(String path) {
        // String rootPath = rootAxFile.getPath();
        //
        // // 앞뒤의 "/"만 제외해준다.
        // if(path.charAt(0) == File.separatorChar) path = path.substring(1);
        // if(path.charAt(path.length() - 1) == File.separatorChar) path =
        // path.substring(path.length() - 1);
        //
        // if(!path.startsWith(rootPath)) return null;
        // int index = rootPath.length();
        // if(path.charAt(index) != File.separatorChar) return null;
        // String subPath = path.substring(index);
        //
        // return new File(rootDirectory, subPath).getAbsolutePath();

        // The path is a relative path.
        return new File(rootDirectory, path).getAbsolutePath();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public String toVirtualPath(String path) {

        // String rootPath = null;
        // try {
        // // 항상 "/"가 제일 앞에 오는 형태로 반환하므로 이를 제외해 준다.
        // rootPath = rootDirectory.getAbsolutePath().substring(1);
        // } catch(IndexOutOfBoundsException e) {
        // // IndexOutOfBoundsException가 발생한다면 잘못된 File.
        // return null;
        // }
        //
        // // 앞뒤의 "/"만 제외해준다.
        // if(path.charAt(0) == File.separatorChar) path = path.substring(1);
        // if(path.charAt(path.length() - 1) == File.separatorChar) path =
        // path.substring(path.length() - 1);
        //
        // if(path.startsWith(rootPath)) return
        // path.substring(rootPath.length());
        //
        // return null;

        // The path is a native full path.
        String nativeRootPath = rootDirectory.getAbsolutePath().substring(1);

        // 앞뒤의 "/"만 제외해준다.
        try {
            if (nativeRootPath.charAt(0) == File.separatorChar)
                nativeRootPath = nativeRootPath.substring(1);
            if (nativeRootPath.charAt(nativeRootPath.length() - 1) == File.separatorChar)
                nativeRootPath = nativeRootPath.substring(0, nativeRootPath.length() - 1);
        }
        catch (IndexOutOfBoundsException e) {
            return null;
        }

        // 앞뒤의 "/"만 제외해준다.
        try {
            if (path.charAt(0) == File.separatorChar) path = path.substring(1);
            if (path.charAt(path.length() - 1) == File.separatorChar)
                path = path.substring(0, path.length() - 1);
        }
        catch (IndexOutOfBoundsException e) {
            return null;
        }

        String rootPath = rootAxFile.getPath();
        if (path.startsWith(nativeRootPath))
            return rootPath + path.substring(nativeRootPath.length());
        return null;
    }

    /**
     * 파일의 읽기가능 여부를 설정.
     * 
     * @param canRead 읽기가 가능하도록 설정하려면 {@literal true}, 아니면 {@literal false}
     * @since 1.0
     */
    protected void setRead(boolean canRead) {
        this.canRead = canRead;
    }

    /**
     * 파일의 쓰기가능 여부를 설정.
     * 
     * @param canWrite 쓰기가 가능하도록 설정하려면 {@literal true}, 아니면 {@literal false}
     * @since 1.0
     */
    protected void setWrite(boolean canWrite) {
        this.canWrite = canWrite;
    }

    /**
     * File에 대한 DefaultFile을 생성하여 반환.
     * 
     * @param peer File 객체
     * @return File에 대한 DefaultFile
     * @since 1.0
     */
    protected synchronized DefaultFile createDefaultFile(File peer) {
        if (peer.equals(rootAxFile.getPeer())) return (DefaultFile) rootAxFile;

        String cacheKey = peer.getAbsolutePath();
        DefaultFile axFile = cache.get(cacheKey);
        if (null != axFile) return axFile;

        DefaultFile parent = createDefaultFile(peer.getParentFile()); // Warning!
                                                                      // Recursive
                                                                      // call.
        axFile = new DefaultFile(this, parent, peer);

        cache.put(cacheKey, axFile);
        return axFile;
    }
}
